Only "continue" on btn-primary buttons

BryceStevenWilley opened the pull request:
Not btn-warning or btn-danger, which are the default selections for Restart and Exit buttons on kickout screens. Without this, it's easy for a test with a bad `target id` to get into an infinite loop, getting sent to the kickout screen, restarting back to the beginning, and starting again.

All `next` / continue buttons on have the `btn-primary` class. It's not bullet proof, but it's okay? The only difference I could tell is that the `next` buttons don't have an associated `value` attribute, and `restart` does, though I think that's only for a list of buttons. We can't really tell what the buttons will do from those values, but we do know that they aren't continue buttons.

Needs https://github.com/plocket/docassemble-ALAutomatedTestingTests/pull/198 to be successful.

BryceStevenWilley committed:
Only "continue" on btn-primary buttons

Not btn-warning or btn-danger, which are the default selections for Restart
and Exit buttons on kickout screens. Without this, it's easy for a test with
a bad `target id` to get into an infinite loop, getting sent to the kickout
screen, restarting back to the beginning, and starting again.

plocket commented on tests/features/interactive_steps.feature:
```diff
@@ -145,4 +145,19 @@ Scenario: Base64 encoded corner cases are decoded correctly
\ No newline at end of file
+    | checkbox_dict['single_quote_key']['🁳🧴🯍🥫🅍🗎🹇🗷🌲🯵'] | True | |
+
+@fast @i6
+Scenario: Doesn't try to 'continue' with a restart button
```
A failing scenario should start with "Fails with". Also, an explanation as to why this both fails and succeeds might be useful as it's a bit of a weird case.
```suggestion
# This test should fail because it's target id is never found, but that also means
# it has succeeded because it didn't try to keep looking for the id by pressing 'restart'
# to continue.
@fast @i6
Scenario: Fails as it doesn't try to 'continue' with a restart button
```

BryceStevenWilley replied:
TBH the main reason the tests is there is to avoid an infinite loop, not to particularly test that the failure occurs, which IMO makes it different from the other failure tests, i.e. if there was a way for the test to succeed instead, I would do that. But I guess it's better to match the convention, and just explain in the comments as you did.

plocket reviewed (approved):
Great improvement! Very excited for this case to be handled!

Items that I can see for improvement that I couldn't add as comments:
- Adding this to [`continue_exists()`](https://github.com/SuffolkLITLab/ALKiln/blob/7f7bf438fd058c06daddc5c03bd50464d8108e6d/lib/scope.js#L711) as well so such a test could finish faster under these circumstances. Not sure how to add a test for that since I think it's just used in the random input Step right now. Maybe if the YAML was a one-page test it would be more effective, but I'm not sure how a test can be shown to end correctly.
- Updating the CHANGELOG

Maybe we can make an issue for figuring out how to give this a better failure message than the timeout message? I hesitate to alter the default timeout message to be more vague. No elegant solution springs to mind at the moment. Maybe using `continue_exists()` in `setFields()` would be one way.

Regardless, I think this serves the purpose it describes. Let me know if you want me to review again if you change something you want a second eye on.

BryceStevenWilley committed:
Change scenario wording, add comment instead

Co-authored-by: plocket <52798256+plocket@users.noreply.github.com>

BryceStevenWilley committed:
Use new continue selector in continue_exists()

Starts sharing a struct member between continue() and continue_exists().
Also adds to CHANGELOG.

BryceStevenWilley commented:
Added to the change log, and started using the selector in `continue_exists()` by sharing the same selector between the two functions. I'll wait for the new commit to pass tests, and then I'll merge.

> Maybe we can make an issue for figuring out how to give this a better failure message than the timeout message?

I think that's covered by https://github.com/SuffolkLITLab/ALKiln/issues/613, which if implemented would make timeout and not finding the correct interaction on a screen would be different error messages.

plocket commented:
Tests passed and ready to merge?
